### ISSUE TITLE
Support array de-structuring / de-tuping

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#15445]
When doing index-neutral programming, Chapel's support for detupling is invaluable:

```chapel
var (key, val) = kvTuple;
```

This issue requests a similar thing for arrays, perhaps with the option to just pull off a subset of the array's elements rather than having to enumerate all of them.  I'm not sure whether it would be better to express this with tuple or array syntax.  For example, imagine one of the following idioms for grabbing the only or first three elements from an array:

```chapel
var (a1, a2, a3) = myThreeElemArray;
var (b1, b2, b3, ...) = myAtLeastThreeElemArray;
var [c1, c2, c3] = myThreeElemArray;
var [d1, d2, d3, ...] = myAtLeastThreeElemArray;
```

**Comment by e-kayrakli** [comment]
My initial reaction was that "of course it should be using `()` and not `[]`" I think this is because I see the line `var (a1, a2, a3)` as a declaration of tuple of variables (for a loose definition of tuple) and with that mental framework "array of variables" doesn't make much sense to me because array is somewhat a more dynamic concept than tuple.

That being said, I didn't like `var (d1, d2, d3, ...)` very much. I am wondering whether slices can be used for this. It would have the additional benefit of stripping every other element and other acrobatics we can pull with them.

```chapel
var (b1, b2, b3) = myAtLeastThreeElemArray[0..2];
var (even1, even2) = myAtLeastThreeElemArray[0..2 by 2];  // get the 0th and 2th index only
var (first, last) = myAtLeastThreeElemArray[0..len by len]; // maybe too much?
```

This has the additional benefit of not requiring parser changes.

**Comment by ben-albrecht** [comment]
Similarly, it'd also be nice to have a way to unroll an array into arguments:

```chpl
var A = [1, 2, 3];

proc f(a, b, c) {
  return a + b +c;
}

var sum = f(...A); // or some other syntax

**Comment by bradcray** [comment]
> I am wondering whether slices can be used for this.

I agree they could (and don't think anything would prevent your examples from working if we implemented the `...` proposal), but I think a benefit of having some sort of "and then drop everything else on the floor" syntax is that it lets someone grab the prefix of a variable-length array without the runtime overhead of slicing.

I suppose we could put in compiler effort to optimize the slice away in some / all cases, but I think I'd much rather adjust the front-end to parse `...` and know that it means "the array's size doesn't have to match" than to have to implement that analysis and optimization.

**Comment by bradcray** [comment]
> Similarly, it'd also be nice to have a way to unroll an array into arguments:

I agree that this would be conceptually attractive, but it's much more challenging than the feature proposed here since we don't currently tend to know the sizes of arrays statically (even when they appear obvious as in your example), which would be required to implement your feature (whereas the de-tupling pattern indicates the size of the array [prefix] in my feature).  Feel free to fork off an issue requesting it, though.

**Comment by ben-albrecht** [comment]
>  I agree that this would be conceptually attractive, but it's much more challenging than the feature proposed here since we don't currently tend to know the sizes of arrays statically.

Agreed.

> Feel free to fork off an issue requesting it, though.

Opened here: https://github.com/chapel-lang/chapel/issues/15518

**Comment by e-kayrakli** [comment]
> I agree they could (and don't think anything would prevent your examples from working if we implemented the ... proposal), but I think a benefit of having some sort of "and then drop everything else on the floor" syntax is that it lets someone grab the prefix of a variable-length array without the runtime overhead of slicing.
>
> I suppose we could put in compiler effort to optimize the slice away in some / all cases, but I think I'd much rather adjust the front-end to parse ... and know that it means "the array's size doesn't have to match" than to have to implement that analysis and optimization.

Sounds reasonable to me. I can live with that syntax because of these benefits.
